### ISSUE TITLE
Python 3.5 and 2.7

### ISSUE THREAD

**Issue by lifeguard999** [dsuch/pymqi#60]
Hello,

I noticed @sky-chen has done some work in this regard a while back (#13) and I understand you're happy to go ahead with Python 3 as long as backward compatibility with 2.7 is retained (as per #40).
If you can confirm this is still the case and you don't know of anyone else starting this initiative already (I assume #40 didn't work out in the end), I'd be happy to make the necessary changes to support py2.7 and py3.5 (and later). I'll be joining a 24h hackathon on 21/22 April and if this project is accepted we'll be doing this in a team 4-people strong.

My motivation: I found myself having to maintain another py3 fork of your library and frankly I'd rather spend some personal time to make yours better than having to maintain that fork.

**Comment by dsuch** [comment]
I'm not aware of any works related to Python 3 compatibility but note that such a port also needs to:

* Include support for Python 2.6 onwards rather than 2.7+ (because RHEL 6 uses Python 2.6 by default)
* Compile correctly, without warnings, under MSVC, but here it is fine to assume 2.7+

There are various GUI products embedding PyMQI and Python 2.1-2.2 that rarely get changed which is why currently PyMQI supports Python 2.1-2.7.

Going forward we can assume 2.6 or greater but 2.7+ is not an option for at least several years until RedHat stop to support RHEL 6.

**Comment by lifeguard999** [comment]
Thanks for clarifying.
Btw, how do you feel about adding dependency on a compatibility package such as six? I'm not sure yet if it'll be of any help, but need to know whether it's a possibility.

**Comment by dsuch** [comment]
Unfortunately, having any dependencies in this project is not really feasible. The exception are dependencies for testing but they cannot be required by the main functionality to install.

PyMQI has to run in all sorts of enterprise environments, including niche ones such as HP-UX, very often without Internet connectivity. Moreover, MQ projects tend to live for 10-20 years, I still encounter MQ 6.x or 7.x queue managers doing their jobs, particularly on z/OS.

But, PyMQI is not a very big project and there will be likely just a few functions needed from six so they can be simply inlined, i.e. embedded into pymqi.py - most of what six does is really straightforward so it should not be any issue.

An alternative that I don't want to employ is to embed six itself in PyMQI but this is ungainly in the long run - in Zato we have 150+ dependencies, some of them ship with their own copy of six, each in a different version, and sorting out subtle incompatibilities between them is a major hassle.

**Comment by lifeguard999** [comment]
Just an FYI, started working on this now. Your notes about Py 2.6 and no dependencies taken on board. I think I have CMQC.py now, it was quite short. But in fact tricky! Suggested solution is:
```python
# Python 2 only
bigint = 1L

# Python 2 and 3
from builtins import int
bigint = int(1)
```

[http://python-future.org/compatible_idioms.html#long-integers]

but "builtins" is only there after pip installing "future", so I reckon you'd rather avoid that.
Thus after reading a bit about how ints and longs work, and trying to avoid excessive usage of ifs, I came with this home-baked solution:
```python
def py23long(x):
    return x + 0 * 0xffffffffffffffff
bigint = py23long(1)
```
That forces python 2 to convert int to long (if it isn't already). Also works in py3 and doesn't rise IDE warnings in neither.

**Comment by lifeguard999** [comment]
Need your opinion on string handling, got stuck on Python 3 implementation of 'get' function. 
MQ libs use 'bytes' (i.e. char* in C) behind the scenes to send and receive information. Conveniently, in Python 2 bytes and strings are one and the same, which gives a nice outcome, where one can send & receive binary data the same way they'd send and receive a 'hello world' text.
Python 3 breaks that compatibility; 'str' and 'bytes' are no longer the same. Moreover, the default is 'str' (and it means only str, i.e. it doesn't seem to accept arbitrary binary data any more) which means every interaction with MQ libs will result in some sort of conversion to/from 'bytes'. To add pain to the business, the conversion is not 100% reliable, especially when converting from 'bytes' to 'str' (i.e. when receiving, i.e. in 'get'). In Python 3 'str' will no longer accept values which are not a valid text (ASCII or unicode I guess), so converting a binary data to 'str' will fail if receiving binary data.
All that leads me to conclude there's no reliable way to implement a single 'get' function in Python 3. My proposal is to implement extra 2 functions which will function in both Py2 and Py3: 'get_bytes' and 'get_str'. But an old-style 'get' can only work in Py2 and will fail in Py3.

E.g.:
```python
# Python 2
put('text')
put(b'bytes')
s = get()  # s is instance of both 'str' and 'bytes' in python 2

# Python 3
put_str('text')
put_bytes(b'text')
s = get_str()    # s is instance of str
b = get_bytes()  # b is instance of bytes
```

Background: in Python 2 a "str" and a "byte" seems to be one and the same, i.e.:
```python
# Python 2: all return True:
isinstance('Hello World', str)     # True
isinstance('Hello World', bytes)   # True
isinstance(b'Hello World', str)    # True
isinstance(b'Hello World', bytes)  # True

# Python 3: bytes and str got separated:
isinstance('Hello World', str)     # True
isinstance('Hello World', bytes)   # False
isinstance(b'Hello World', str)    # False
isinstance(b'Hello World', bytes)  # True
```

To convert from 'str' to 'bytes' Python 3 one needs to explicitly state encoding (everyone seems to go for 'utf-8') and this seems to work reliably most of the time.
To convert from 'bytes' to 'str' one needs the same (again, utf-8 does the job), but fails when data is binary, not text.

Thoughts welcome...

**Comment by dsuch** [comment]
OK, the idea with two functions is to make it explicit for users to choose what they want to receive on output, is that correct?

In other words, if a Java application sends a message there would be no need for any sort of foreknowledge on a PyMQI user's end whether to use .get_str or .get_bytes and both could be used, depending on one's needs to receive str or bytes?

If so, I'm still not clear why .get in Python 3 cannot default to something that we arbitrarily settle on, or can it?

**Comment by lifeguard999** [comment]
Correct, I was proposing 2 functions only to distinguish return result.

The answer to Java analogy is also yes, one would choose a function depending on how one chooses to parse the message (bytes == raw, str == parsed from bytes to str).

Actually I like the idea of deciding on single get() instead. I have impression bytes will be more suited for the job for several reasons - firstly this is what actually gets sent over TCP/IP and what IBM's API returns so there will be less parsing by default, second Python 3 'str' doesn't seem to play well with binary data (e.g. if one receives a zip-compressed data) so 'bytes' will work in generic case of receiving any data (be that raw/compressed/pdf as well as text/csv/xml/json).

The impact of this decision on py3 users is they'll need to make sure they treat the bytes right, e.g.:
```python
# Python 2
data = get(...)
assert data == '\x12\x34\x23'  # data works as expected

text = get(...)
assert text == 'Hello!'  # strings work as expected

# Python 3
data = get(...)
assert data == b'\x12\x34\x23'  # data works as expected

text = get(...)
assert text == 'Hello!'  # this will no longer work!!!
assert text == b'Hello!'  # one needs to use byte-strings (with b prefix)
assert text.decode('ascii') == 'Hello!'  # or convert bytes to string by specifying encoding.
```

**Comment by dsuch** [comment]
Thanks, yes, it's best to assume that data taken off queues is bytes instead of string.

**Comment by lifeguard999** [comment]
Looks like I have a version passing all nosetests & release.py on Python 3.5 (theoretically back-compatible with 2.6). Will move onto re-testing on 2.6 next.
A prototype diff if you'd like to give some early comments: 
https://github.com/dsuch/pymqi/compare/master...lifeguard999:master
(note quite a few differences are to sort out PyCharm warnings, I plan to take them out later, but for the time being it helps me find issues faster).

**Comment by dsuch** [comment]
Thanks @lifeguard999 - I will have a look at everything this week, though I cannot confirm yet when exactly.

**Comment by dsuch** [comment]
Hello @lifeguard999,

I think it all looks good - I don't have any specific comments short of confirming that testfixtures can be removed from the requirements and that it was a good idea to rename things from camelCase to snake_style or to replace references to apply in __init__ methods - the original implementation was prepared some 15 years ago and it showed.

Thanks.

**Comment by lifeguard999** [comment]
Apologies for delay, but now back on it. Tests in the pipeline on Win 10 (64 bit):
* Python 2.7 + "Microsoft Visual C++ Compiler for Python 2.7.x"
pass
(Note because I changed setuptools to distutils, there's a problem to overcome in this particular setup, as per https://github.com/cython/cython/wiki/CythonExtensionsOnWindows)
* Python 3.5 + MSVC 2015
pending...
* Python 2.6 + MSVC 2008
pending...

I can't quite see an easy way to test with 32bit system though, it appears MQ client demo only comes in 64 bit version. Advise welcome.

**Comment by lifeguard999** [comment]
Dug a bit deeper into 32 bit clients.
Apparently MQ version 9 only comes with 64 bit support (https://developer.ibm.com/answers/questions/293315/is-there-a-32-bit-release-for-ibm-mq-90/).
At the same time I was not able to find a download link for trial of version 8.

Meanwhile however, managed to run partial tests on 32bit Python 2.7 in Win; this was in a largely restricted set up however, thus there was no authorisation to run more elaborate tests like creating queues, topics or setting up parameters.
Nonetheless, connecting, sending and receiving passed. RFH2 tests also passed. Everything else failed with MQRC_NOT_AUTHORIZED.

**Comment by JoshMcMQ** [comment]
@lifeguard999 I added my own comment to the developer.ibm.com link you provided.  MQ v9 does not come with a package that is specifically for 32bit hardware, but the 64bit package does contain 32bit libraries to allow 32 bit apps to be compiled and work with the MQ client.  The same is true of v8 only it does provide a package specifically for 32bit hardware.  The download links for MQ client can be found searching on google for either of the following terms (Note each term below is also a link to the IBM page that can be found in the google search):
- [MQC8](https://www.ibm.com/support/docview.wss?uid=swg24037500)
- [MQC9](https://www.ibm.com/support/docview.wss?uid=swg24042176)

**Comment by dsuch** [comment]
Thanks for the update on the Windows works @lifeguard999 - this sounds very nice indeed. As for the 32-bit client, if you can test it out in addition to the 64-bit, that would be great.

**Comment by hugovk** [comment]
I wouldn't worry too much about Python 2.6/RHEL 6.

Here's the pip installs for pymqi from PyPI for April 2018:

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 2.7            |  84.48% |            294 |
| 3.6            |   9.77% |             34 |
| 3.5            |   2.87% |             10 |
| 3.4            |   1.72% |              6 |
| 2.6            |   0.57% |              2 |
| 3.3            |   0.29% |              1 |
| 3.7            |   0.29% |              1 |
| Total          |         |            348 |

Source: `pypinfo --start-date 2018-04-01 --end-date 2018-04-30 --percent --pip --markdown pymqi pyversion`

And distributions:

| distro_name                     | distro_version | percent | download_count |
| ------------------------------- | -------------: | ------: | -------------: |
| Ubuntu                          |          17.10 |  51.95% |            173 |
| None                            |           None |  17.72% |             59 |
| Ubuntu                          |          16.04 |  15.62% |             52 |
| Debian GNU/Linux                |              9 |  10.51% |             35 |
| CentOS Linux                    |              7 |   1.20% |              4 |
| Red Hat Enterprise Linux Server |            7.5 |   0.60% |              2 |
| Raspbian GNU/Linux              |              9 |   0.60% |              2 |
| Red Hat Enterprise Linux Server |            7.2 |   0.60% |              2 |
| macOS                           |        10.13.4 |   0.60% |              2 |
| Debian GNU/Linux                |              8 |   0.60% |              2 |
| Total                           |                |         |            333 |

Source: pypinfo --start-date 2018-04-01 --end-date 2018-04-30 --percent --pip --markdown pymqi distro distro-version

**Comment by dsuch** [comment]
Thanks @hugovk - I was not aware of this tool, I will have a closer look.

One truly puzzling element of it is Ubuntu 17.10 constituting more than a half of all the results. What could that be? Someone's automated builds?

**Comment by lifeguard999** [comment]
Followed advice from @JoshMcMQ and it worked! Thanks. 
Got some C compile warnings about int loss of accuracy on 32 bit, so corrected that and re-run all the test configs. Details of test envs and results below.

__Tests performed__
Unless otherwise stated, a 'pass' indicates successful run of both the 'release.py' file (which checks simple connecting/sending/receiving) and running 'nosetests' command which triggers the test pack (sending, receiving, running MQ commands, topics, RFH2 headers and others).
The test packs are mostly unaffected by this change, apart from occasional conversion from print 'abc' to print('abc').

__Test results__

 . | Linux (64bit) | Win (64bit)
--- | --- | ---
Py2.6 (32 bit) |  pass(*1) | pass(*1)
Py2.6 (64 bit) |  pass(*1) | n/a(*2)
Py2.7 (32 bit) |  pass | pass
Py2.7 (64 bit) |  pass | pass
Py3.5 (32 bit) |  pass | pass
Py3.5 (64 bit) |  pass | pass

(*1): release.py - pass, nosetests - mostly pass. 3 cases fail due to incompatibilities in test pack with Py 2.6. They also fail the same way in current live pymqi version.
(*2): Did not manage to compile 64 bit version

__Software__
* Linux (dev & test OS): CentOS 7-1708 VM, 64 bit, downloaded from osboxes.org
  * compiler: gcc (built-in)
  * 32 bit support:
     yum -y install glibc.i686
     yum install glibc-devel.i686 libstdc++-devel.i686 ncurses-devel.i686
     yum install libgcc.i686
     (to build, prepend with CFLAGS, as so)
     CFLAGS=-m32 python setup.py build_ext --inplace
* Win (test OS): Windows 10 VM, 64 bit, trial downloaded from microsoft.com
   * compilers:
       Py2.6 - MSVC++ 2008 Express (not easily found 10 years on, but there was a link on some forum pointing to microsoft.com download)
               Comes with 32 bit command line by default
       Py2.7 - MSVC++ Compiler Package for Python 2.7 (available on microsoft.com)
               Comes with 32 bit and 64 bit command lines
       Py3.5 - MSVC++ 2015 Community Edition (available on microsoft.com)
               Comes with 32 bit and 64 bit command lines
* IBM MQ server & client: 
  * Linux 64 (for Redhat-like, works with CentOS): v9.0.0.0, 64 bit, trial downloaded from ibm.com
  * Win 64: v9.0.0.2, 64 bit, trial downloaded from ibm.com
* Python: Base of Anaconda with Python 2.7 64 bit, with following virtual envs:
        (Each virtual env only contains minimum Python packages, plus manual pip install of 'testfixtures' and 'nose' to run tests)
    Py26 (64 bit) (2.6.9)
    Py27 (64 bit) (2.7.14 Linux, 2.7.15 Win)
    Py35 (64 bit) (3.5.5)
    Py26_32 (32 bit) (2.6.9)
    Py27_32 (32 bit) (2.7.15)
    Py35_32 (32 bit) (3.5.5)

**Comment by dsuch** [comment]
Thanks @lifeguard999 - this looks very nice, I think you are ready to send a PR in that case?

**Comment by lifeguard999** [comment]
@dsuch - PR raised.

**Comment by dsuch** [comment]
Thanks, I have just merged it - this was good stuff. I will be releasing a new version to PyPI soon.

**Comment by JoshMcMQ** [comment]
@dsuch Note I found this other project called py3mqi that someone forked from you in September 2017: https://github.com/Dynatrace-Reinhard-Pilz/py3mqi/

I'll put a issue on that project linking to this issue.

**Comment by dsuch** [comment]
Thanks @JoshMcMQ - that sounds great, thanks.

We are putting out a new [Zato](https://zato.io/docs/) release this week and I'm completely tied up with the preparations but right after it I will publish a new PyMQI version with all the changes by @lifeguard999 to whom I can again I am very grateful.

**Comment by dsuch** [comment]
Hello,

I have just released PyMQI 1.8 which includes support for Python 3.5 as implemented by @lifeguard999 - thanks again for that!
